File: f77/build07/kmtc.f
C
        Subroutine Kinematics_Data
C       This subroutine sets the default data for the kinematics routine.
C
C       Inputs:
C           None.
C
C       Outputs:
C           Theta_b_IC_dg - Initial attitude of missile [deg]
C           Velocity - Velocity of missile [m/sec]
C           X_bi_i_IC - Initial X position of BCS w.r t ICS, expressed in the
C                                   ICS [m]
C           Z_bi_i_IC - Initial Z position of BCS w.r t ICS, expressed in the
C                                   ICS [m]
C
C       Internal variables and constants:
C           None.
C
C       Declare global common and assign variable locations
C
        Include '../driver/global.inc'
        Include '../driver/sysvars.inc'
        Equivalence (Real_Array(508), Theta_b_IC_dg)
        Equivalence (Real_Array(510), Velocity)
        Equivalence (Real_Array(506), X_bi_i_IC)
        Equivalence (Real_Array(507), Z_bi_i_IC)
C
C       Begin default data definitions:
C
        Theta_b_IC_dg =    0.0e0
        Velocity      =  100.0e0
        X_bi_i_IC     = -500.0e0
        Z_bi_i_IC     = -100.0e0
C
        Return
        End
C
        Subroutine Kinematics_Init
C
C       This subroutine initializes the variables for the kinematics routine
C
C       Inputs:
C           RDTODG - Radians to degrees conversion factor [deg/rad]
C           Theta_b_IC_dg - Initial attitude of missile [deg]
C           X_bi_i_IC - Initial X position of BCS w.r t ICS, expressed in the
C                                   ICS [m]
C           Z_pi_i_IC - Initial Z position of BCS w.r t ICS, expressed in the
C                                   ICS [m]
C
C       Outputs:
C           Theta_b - Attitude of missile [rad]
C           X_bi_i - X position of BCS w.r.t. ICS, expressed in the ICS [m]
C           Z_bi_i - Z position of BCS w.r.t. ICS, expressed in the ICS [m]
C
C       Internal variables and constants:
C           None.
C
C       Declare global common and assign variable locations
C
        Include '../driver/global.inc'
        Include '../driver/sysvars.inc'
        Equivalence (Real_Array( 50), RDTODG)
        Equivalence (Real_Array(504), Theta_b)
        Equivalence (Real_Array(500), X_bi_i)
        Equivalence (Real_Array(501), Z_bi_i)
        Equivalence (Real_Array(508), Theta_b_IC_dg)
        Equivalence (Real_Array(506), X_bi_i_IC)
        Equivalence (Real_Array(507), Z_bi_i_IC)
C
C       Begin math model initialization:
C
C       Define the states
C
        Call Define_Real_State(500, 502)
        Call Define_Real_State(501, 503)
        Call Define_Real_State(504, 505)
C
C       Set the initial conditions
C
        X_bi_i = X_bi_i_IC
        Z_bi_i = Z_bi_i_IC
        Theta_b = Theta_b_IC_dg/RDTODG
C
        Return
        End

        Subroutine Kinematics
C
C       This subroutine determines the kinematics state derivatives
C
C       Inputs:
C               Theta_b - Attitude of missile [rad]
C               Velocity - Velocity of missile [m/sec]
C
C       Outputs:
C               Xd_bi_i - X velocity of BCS w.r.t. ICS, expressed in the ICS [m/sec]
C               Zd_bi_i - Z velocity of BCS w.r.t. ICS, expressed in the ICS [m/sec]
C
C       Internal variables and constants:
C
C       Declare global common and assign variable locations
C
        Include '../driver/global.inc'
        Include '../driver/sysvars.inc'
        Equivalence (Real_Array(510), Velocity)
        Equivalence (Real_Array(502), Xd_bi_i)
        Equivalence (Real_Array(503), Zd_bi_i)
        Equivalence (Real_Array(504), Theta_b)
C
C       Begin math model:
C
C       Calculate velocities (attitude rate comes directly from airframe response)
        Xd_bi_i = Velocity*Cos(Theta_b)
        Zd_bi_i = -Velocity*Sin(Theta_b)
C
        Return
        End

File: gcc/build07/kmtc.c
#include "kmtc.h"
#include "../driver/global.h"
#include "../driver/sysvars.h"
#include "../driver/deriv.h"
#include "../driver/defstate.h"
#include <math.h>

#define rdtodg        real_array[50].r  // Radians to degrees conversion factor [deg/rad]
#define x_bi_i        real_array[500].r // X position of BCS w.r.t. ICS, expressed in the ICS [m]
#define z_bi_i        real_array[501].r // Z position of BCS w.r.t. ICS, expressed in the ICS [m]
#define xd_bi_i       real_array[502].r // X velocity of BCS w.r.t. ICS, expressed in the ICS [m/sec]
#define zd_bi_i       real_array[503].r // Z velocity of BCS w.r.t. ICS, expressed in the ICS [m/sec]
#define theta_b       real_array[504].r // Attitude of missile [rad]
#define x_bi_i_ic     real_array[506].r // Initial X position of BCS w.r t ICS, expressed in the ICS [m]
#define z_bi_i_ic     real_array[507].r // Initial Z position of BCS w.r t ICS, expressed in the ICS [m]
#define theta_b_ic_dg real_array[508].r // Initial attitude of missile [deg]
#define velocity      real_array[510].r // Velocity of missile [m/sec]

void kinematics_data() {
  // This subroutine sets the default data for the kinematics routine.
  //
  // Inputs:
  //     None.
  //
  // Outputs:
  //     Theta_b_IC_dg - Initial attitude of missile [deg]
  //     Velocity - Velocity of missile [m/sec]
  //     X_bi_i_IC - Initial X position of BCS w.r t ICS, expressed in the ICS [m]
  //     Z_bi_i_IC - Initial Z position of BCS w.r t ICS, expressed in the ICS [m]
  // Begin default data definitions:
  theta_b_ic_dg = 0.0;
  velocity = 100.0;
  x_bi_i_ic = -500.0;
  z_bi_i_ic = -100.0;
  }

void kinematics_init() {
  // This subroutine initializes the variables for the kinematics routine
  //
  // Inputs:
  //     RDTODG - Radians to degrees conversion factor [deg/rad]
  //     Theta_b_IC_dg - Initial attitude of missile [deg]
  //     X_bi_i_IC - Initial X position of BCS w.r t ICS, expressed in the ICS [m]
  //     Z_pi_i_IC - Initial Z position of BCS w.r t ICS, expressed in the ICS [m]
  //
  // Outputs:
  //     Theta_b - Attitude of missile [rad]
  //     X_bi_i - X position of BCS w.r.t. ICS, expressed in the ICS [m]
  //     Z_bi_i - Z position of BCS w.r.t. ICS, expressed in the ICS [m]
  // Begin math model initialization:
  // Define the states
  define_real_state(500,502);
  define_real_state(501,503);
  define_real_state(504,505);
  // Set the initial conditions
  x_bi_i = x_bi_i_ic;
  z_bi_i = z_bi_i_ic;
  theta_b = theta_b_ic_dg/rdtodg;
}

void kinematics() {
  // This subroutine determines the kinematics state derivatives
  //
  // Inputs:
  //     Theta_b - Attitude of missile [rad]
  //     Velocity - Velocity of missile [m/sec]
  //
  // Outputs:
  //     Xd_bi_i - X velocity of BCS w.r.t. ICS, expressed in the ICS [m/sec]
  //     Zd_bi_i - Z velocity of BCS w.r.t. ICS, expressed in the ICS [m/sec]
  //
  // Internal variables and constants:
  //
  // Declare global common and assign variable locations
  // Begin math model:
  // Calculate velocities (attitude rate comes directly from airframe response)
  xd_bi_i =  velocity*cos(theta_b);
  zd_bi_i = -velocity*sin(theta_b);
}
